### DISCUSSION TITLE
Indexing daemon model

### DISCUSSION THREAD

**Discussion by cyphercider** [vaultec81/3Speak-research#2] [High-Level Designs]
# Indexing daemon model

For discussing the index cache model used by SPK indexing nodes

### Responsibilities

Caching of full document content from ceramic for rapid local retrieval 

### Open questions

1. Can we build verification of cryptographic signatures of document contents into our indexing process, to verify that the claimed publisher of the content actually signed the content?

### Draft definition

```ts
export class GraphDocument implements SpkCollectionItem {
  // Mongo internal ID
  _id: ObjectId
  /**
   * Stream ID
   */
  id: string
  /**
   * Parent stream id
   */
  parentId?: string
  content: unknown
  expire?: any
  schemaStreamId: string
  updated_at: Date
  last_checked: Date
}
```

### Responsibilities

1:1 with `IndexedDocument`.  Stores graph management related metadata for each indexed document.

### Open Questions

1. Is the custodianNodes array on this type duplicative with the same property on the Subgraph, and should it be removed?
2. What level at which do we want to allow custodian nodes to announce custody of a subgraph?  Are subgraphs small enough that a custodian node must commit to custody an entire subgraph rather than allowing them to custody only part of a subgraph?

### Draft definition

```ts
export class GraphIndex implements SpkCollectionItem {
  _id: ObjectId
  // Stream ID
  id: string
  children: string[]
  expiration: any
  custodianNodes: any[] // TODO - figure out the correct type for a custodian node
}
```

### Responsibilities

Tracking of each subgraph at a high level.  Keeps track of which foreign custodian nodes are tracking the subgraph.  

### Open Questions

1. Do we like this new collection to track each subgraph?
2. Do we like the idea of tracking a list of valid schemas for a subgraph?
3. What is the format of the custodian node datatype?  Is it just an IP address?

### Draft definition

```ts
export class CustodianMap implements SpkCollectionItem {
  _id: ObjectId
  /**
   * The stream ID of the root node of the subgraph.  Also the subgraph ID.
   */
  id: string
  /**
   * list of valid schemas for the subgraph by stream ID
   */
  validSchemas: string[]
  /**
   * List of custodian nodes that have announced they are storing this subgraph
   */
  custodianNodes: any[]
}
```

**Comment by vaultec81** [comment]
> Can we build verification of cryptographic signatures of document contents into our indexing process, to verify that the claimed publisher of the content actually signed the content?

This should be handled by Ceramic already. No need to do extra verification unless we have a specific reason to (like checking against blockchain records)

**Comment by vaultec81** [comment]
> Do we like this new collection to track each subgraph?

I think `graph.index` collection should be enough to do all the parent-child tracking we'd need.

> What is the format of the custodian node datatype? Is it just an IP address?

The Libp2p PeerId. This might change in the future to something we have more control over or is more neutral, but in nearly all cases the peerId is enough to do everything we need. If we need to store/keep track of other metadata we can do that in a `peers` collection or something of that nature.
